libs/langchain-azure-openai/src/chat_models.ts

```diff
@@ -0,0 +1,687 @@
+    });
+  } else if (role === "tool") {
+    return new ToolMessageChunk({
+      content,
+      additional_kwargs,
+      tool_call_id: delta.toolCalls[0].id,
+    });
+  } else {
+    return new ChatMessageChunk({ content, role });
+  }
+}
+
+function openAIResponseToChatMessage(
+  message: ChatResponseMessage
+): BaseMessage {
+  switch (message.role) {
+    case "assistant":
+      return new AIMessage(message.content || "", {
+        function_call: message.functionCall,
+        tool_calls: message.toolCalls,
+      });
+    default:
+      return new ChatMessage(message.content || "", message.role ?? "unknown");
+  }
+}
+
+interface OpenAILLMOutput {
+  tokenUsage: TokenUsage;
+}
+
+function extractGenericMessageCustomRole(message: ChatMessage) {
+  if (
+    message.role !== "system" &&
+    message.role !== "assistant" &&
+    message.role !== "user" &&
+    message.role !== "function" &&
+    message.role !== "tool"
+  ) {
+    console.warn(`Unknown message role: ${message.role}`);
+  }
+
+  return message.role;
+}
+
+export function messageToOpenAIRole(message: BaseMessage): string {
+  const type = message._getType();
+  switch (type) {
+    case "system":
+      return "system";
+    case "ai":
+      return "assistant";
+    case "human":
+      return "user";
+    case "function":
+      return "function";
+    case "tool":
+      return "tool";
+    case "generic": {
+      if (!ChatMessage.isInstance(message))
+        throw new Error("Invalid generic chat message");
+      return extractGenericMessageCustomRole(message);
+    }
+    default:
+      throw new Error(`Unknown message type: ${type}`);
+  }
+}
+
+export interface ChatOpenAICallOptions
+  extends OpenAIChatCallOptions,
+    BaseFunctionCallOptions {
+  tools?: ChatCompletionsToolDefinition[];
+  tool_choice?: ChatCompletionsNamedToolSelection;
+  response_format?: ChatCompletionsResponseFormat;
+  seed?: number;
+}
+
+export class AzureChatOpenAI
+  extends BaseChatModel<ChatOpenAICallOptions>
+  implements OpenAIChatInput, AzureOpenAIInput
+{
+  static lc_name() {
+    return "AzureChatOpenAI";
+  }
+
+  get callKeys() {
+    return [
+      ...super.callKeys,
+      "options",
+      "function_call",
+      "functions",
+      "tools",
+      "tool_choice",
+      "promptIndex",
+      "response_format",
+      "seed",
+    ];
+  }
+
+  get lc_secrets(): { [key: string]: string } | undefined {
+    return {
+      azureOpenAIApiKey: "AZURE_OPENAI_API_KEY",
+      azureOpenAIEndpoint: "AZURE_OPENAI_API_ENDPOINT",
+      azureOpenAIApiDeploymentName: "AZURE_OPENAI_API_DEPLOYMENT_NAME",
+    };
+  }
+
+  get lc_aliases(): Record<string, string> {
+    return {
+      modelName: "model",
+      azureOpenAIApiKey: "azure_openai_api_key",
+      azureOpenAIEndpoint: "azure_openai_api_endpoint",
+      azureOpenAIApiDeploymentName: "azure_openai_api_deployment_name",
+    };
+  }
+
+  lc_serializable = true;
+
+  functions?: FunctionDefinition[] | undefined;
```

Comment:
Can you use the function calling types here?

https://github.com/langchain-ai/langchainjs/blob/main/langchain-core/src/language_models/base.ts#L198

They should be equivalent.

Reply:
Yes. They are interoperable. But this section has been removed based on the PR Comment: https://github.com/langchain-ai/langchainjs/pull/3919/#discussion_r1468116004